utils/post_utils.jsx

```diff
@@ -511,31 +512,50 @@ export function getValidUserPostFromId(state, id) {
     return getPost(state, id);
 }
 
+function isPostHeaderVisible(post, previousPost, getReplyCount) {
+    //If there is no previous post then current post must have visible header
+    if (!previousPost) {
+        return true;
+    }
+
+    //If current post is first message of a thread then it must have header visible
+    if (isFirstReply(post, previousPost)) {
+        return true;
+    }
+
+    //If current post is a comment and not the first comment of a thread then its header must be hidden
+    if (post.root_id) {
+        return false;
+    }
+
+    //If current post and previous posts both are not comments
+    if (!post.root_id && !previousPost.root_id) {
+        //If a non-comment post has some replies then its header must be visible
+        //If a non-comment post is a consecutive post then its header must be hidden
+        return getReplyCount(post) > 0 || !areConsecutivePostsBySameUser(post, previousPost);
+    }
+
+    return true;
+}
+
 export function getCurrentUserLastPostGroupFirstPostId(state, postListIds) {
     //This function returns the first post id of the last post group by the current user
-    let currentUserLastPost;
+    const getReplyCount = makeGetReplyCount();
+
+    //This function is made to avoid passing the state to the isPostHeaderVisible function
+    const getReplyCountModified = (post) => {
+        return getReplyCount(state, post);
+    };
 
+    let nextPost;
     for (let i = 0; i < postListIds.length; i++) {
-        const id = postListIds[i];
-        const post = getValidUserPostFromId(state, id);
-        if (!post) {
-            //if it is an combinedUserActivityPost, return if the currentUserLastPost is already set
-            if (currentUserLastPost) {
-                return currentUserLastPost.id;
-            }
-            continue;
-        }
+        const post = nextPost || getValidUserPostFromId(state, postListIds[i]);
+        nextPost = i + 1 < postListIds.length ? getValidUserPostFromId(state, postListIds[i + 1]) : undefined;
 
-        if (isPostOwner(state, post) && !isSystemMessage(post)) {
-            if (currentUserLastPost && !areConsecutivePostsBySameUser(currentUserLastPost, post)) {
-                //currentUserLastPost is set and this post is not consecutive so the currentUserLastPost is the first post of the last postGroup
-                return currentUserLastPost.id;
-            }
-            currentUserLastPost = post;
-        } else if (currentUserLastPost) {
-            return currentUserLastPost.id;
+        if (isPostOwner(state, post) && !isSystemMessage(post) && isPostHeaderVisible(post, nextPost, getReplyCountModified)) {
+            return post.id;
```

Comment:
Check if something like this would work:

```ts
// This function returns the last post id of the current user's post which has a header visible
export function getLastCurrentUserPostWithHeaderId(state, postListIds) {
    const getReplyCount = (post) => makeGetReplyCount()(state, post);

	const firstPostWithHeaderID = postListIds.find((postID, index) => {
		const post = getValidUserPostFromId(state, postID);
		
		let nextPost;
		if (index + 1 < postListIds.length) {
			const nextPostID = postListIds[index + 1];
			nextPost = getValidUserPostFromId(state, nextPostID);
		}

		return (isPostOwner(state, post) && !isSystemMessage(post) && isPostHeaderVisible(post, nextPost, getReplyCount));
	});

	return firstPostWithHeaderID || '';
}
```

Reply:
In this code, every time getReplyCount is called, makeGetReplyCount() is called. Isn't is better if makeGetReplyCount() is called only once just like it is currently being called.

Reply:
The find approach works as expected but the only drawback is that we have to find two posts from ids in every iteration.... but in the for loop approach we don't have to find nextPost if it's already present.

Reply:
1. @manojosh I've updated the original post as well. I think it's better if we use:

```ts
const getReplyCount = makeGetReplyCount().bind(null, state);
```

2. Calling `getValidUserPostFromId` once or twice is still the same order of operations. So, there is not much optimization. We should prioritize readability and maintainability in this case.